### PR TITLE
Adds tests for suppressed constants in postgres_fdw queries

### PR DESCRIPTION
**Pull request by naisila**
PG15 has suppressed some casts on constants when querying foreign tables.

For example, we can use text to represent a type that's an enum on the remote side.
A comparison on such a column will get shipped as "var = 'foo'::text".
But there's no enum = text operator on the remote side.
If we leave off the explicit cast, the comparison will work.

Test we behave in the same way with a Citus foreign table

Reminder: foreign tables cannot be distributed/reference, can only be Citus local

Relevant PG commit:
https://github.com/postgres/postgres/commit/f8abb0f5e114d8c309239f0faa277b97f696d829

### CODE REVIEW

File: src/test/regress/sql/pg15.sql [hunk 1]
```diff
@@ -799,6 +799,81 @@ CREATE TABLE set_on_default_test_referencing(
...
+SELECT * FROM foreign_table WHERE 'foo' = c1 LIMIT 1;
+SELECT * FROM foreign_table WHERE 'foo' = c1 LIMIT 1;
+
+-- we declared c1 to be text locally, but it's still the same type on
+-- the remote which will balk if we try to do anything incompatible
+-- with that remote type
+SELECT * FROM foreign_table WHERE c1 LIKE 'foo' LIMIT 1; -- ERROR
+SELECT * FROM foreign_table WHERE c1::text LIKE 'foo' LIMIT 1; -- ERROR; cast not pushed down
+
+ALTER FOREIGN TABLE foreign_table ALTER COLUMN c1 TYPE user_enum;
```
Comment (hanefi) [inline src/test/regress/sql/pg15.sql:868]: In the PG test file, they continue testing this foreign table, and hence they revert the change on column data type. However we do not really need that here.

```suggestion
```

File: src/test/regress/sql/pg15.sql [hunk 2]
```diff
@@ -799,6 +799,81 @@ CREATE TABLE set_on_default_test_referencing(
...
+-- Clean up foreign table test
+DROP FOREIGN TABLE foreign_table;
+DROP TABLE foreign_table_test;
+\set VERBOSITY default
+RESET citus.use_citus_managed_tables;
+SELECT 1 FROM citus_remove_node('localhost', :master_port);
+
 -- Clean up
 \set VERBOSITY terse
 SET client_min_messages TO ERROR;
```
Comment (hanefi) [inline src/test/regress/sql/pg15.sql:879]: Several ideas on better cleanup:

1. Drop foreign server. They are not associated with a schema, and we need to drop them manually.
2. Do not delete tables and foreign tables, as they will get deleted anyway along with the schema.
3. Move the cleanup logic to the end, and merge with the current items.

Feel free to use a subset of this ideas, and remember that an empty set is a subset of all sets.

```suggestion
-- Clean up
\set VERBOSITY terse
SET client_min_messages TO ERROR;
RESET citus.use_citus_managed_tables;
SELECT 1 FROM citus_remove_node('localhost', :master_port);
DROP SERVER foreign_server CASCADE;
```

File: src/test/regress/sql/pg15.sql [hunk 3]
```diff
@@ -799,6 +799,81 @@ CREATE TABLE set_on_default_test_referencing(
...
+        FOREIGN DATA WRAPPER postgres_fdw
+        OPTIONS (host 'localhost', port :'master_port', dbname 'regression');
+
+CREATE USER MAPPING FOR CURRENT_USER
+        SERVER foreign_server
+        OPTIONS (user 'postgres');
+
+CREATE FOREIGN TABLE foreign_table (
+        c0 integer NOT NULL,
+        c1 user_enum
```
Comment (hanefi) [inline src/test/regress/sql/pg15.sql:841]: It would make it easier for readers to understand this test if we actually create the foreign table without using the enum here.

It made sense for PG to create the foreign table like this, and later alter it. However it hurts readability here IMO.

```suggestion
        c1 text
```

File: src/test/regress/sql/pg15.sql [hunk 4]
```diff
@@ -799,6 +799,81 @@ CREATE TABLE set_on_default_test_referencing(
...
+)
+        SERVER foreign_server
+        OPTIONS (schema_name 'pg15', table_name 'foreign_table_test');
+
+-- check that the foreign table is a citus local table
+SELECT partmethod, repmodel FROM pg_dist_partition WHERE logicalrelid = 'foreign_table'::regclass ORDER BY logicalrelid;
+
+-- same tests as in the relevant PG commit
+ALTER FOREIGN TABLE foreign_table ALTER COLUMN c1 TYPE text;
+
```
Comment (hanefi) [inline src/test/regress/sql/pg15.sql:851]: I think we can drop these lines if we change the `CREATE FOREIGN TABLE` statement above as I suggested.

**Review by hanefi — COMMENTED** [review]
Nice tests. I shared some ideas on improving readability of this test and cleaning up properly.

While you are at it, you can consider rebasing as well.

File: src/test/regress/sql/pg15.sql [hunk 5]
```diff
@@ -799,6 +799,81 @@ CREATE TABLE set_on_default_test_referencing(
 
 SELECT 1 FROM citus_remove_node('localhost', :master_port);
```
Comment (hanefi) [inline src/test/regress/sql/pg15.sql:800]: How about deleting this line, along with the `citus_add_node` you introduced?

Comment (naisila) [reply src/test/regress/sql/pg15.sql:800]: We need to add the coordinator in order to have a Citus local table. However, the previous test also adds and removes the coordinator. So I will remove the `citus_add_node` that I introduced, along with the `citus_remove_node` from the previous test.

File: src/test/regress/sql/pg15.sql (hunk 2 above)
Comment (naisila) [reply src/test/regress/sql/pg15.sql:879]: Thanks for the suggestions. Definitely I need to drop the foreign server.

I prefer leaving this test's cleanup separate from the whole pg15 test file's cleanup such that it's easier to add new tests at the bottom of the file.

**Review by hanefi — APPROVED** [review]
